### PR TITLE
add PAG support for SD Img2Img

### PR DESCRIPTION
**Pull request by SahilCarterr**
# What does this PR do?
Adds PAG (Perturbed-Attention Guidance) support for SD models (`StableDiffusionPAGImg2ImgPipeline`).Continuation of https://github.com/huggingface/diffusers/pull/7944

Fixes https://github.com/huggingface/diffusers/issues/8710 (partially)

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@yiyixuxu
Anyone in the community is free to review the PR once the tests have passed.

***

Usage

```python
from diffusers import AutoPipelineForImage2Image
from diffusers.utils import load_image, make_image_grid
import torch

# Set device to CUDA for GPU processing
device = "cuda"

# Load the image-to-image pipeline with pre-trained weights
pipe = AutoPipelineForImage2Image.from_pretrained("Lykon/dreamshaper-8", enable_pag=True)
pipe.to(device)

# Load an initial image
init_image = load_image(
    "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/cat.png"
)

# Set a random generator with a specific seed
generator = torch.Generator("cuda").manual_seed(31)

# Define the prompt
prompt = "cat wizard"

# Generate the output image using the pipeline
image = pipe(prompt, image=init_image, generator=generator, pag_scale=0.4).images[0]

# Create a grid of the initial and generated images
make_image_grid([init_image, image], rows=1, cols=2)

```
cat_wizrd

### CODE REVIEW

File: tests/pipelines/pag/test_pag_sd_img2img.py [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+            [0.5986328, 0.52441406, 0.3972168, 0.4741211, 0.34985352, 0.22705078, 0.4128418, 0.2866211, 0.31713867]
+        )
+        print(image_slice.flatten())
+        assert (
+            np.abs(image_slice.flatten() - expected_slice).max() < 1e-3
+        ), f"output is different from expected, {image_slice.flatten()}"
+
+@nightly
+@require_torch_gpu
+class StableDiffusionImg2ImgPipelineNightlyTests(unittest.TestCase):
```
Comment (yiyixuxu) [inline tests/pipelines/pag/test_pag_sd_img2img.py:358]: I think we don't need nightly tests here, one slow test is enough :)

**Review by yiyixuxu — COMMENTED** [review]
thanks! I left some comments
~can we see some outputs too?~ (saw it)!

**Comment by yiyixuxu** [comment]
don't forget to call `make style` and `make fix-copies` too

**Comment by SahilCarterr** [comment]
added "# copied from" , removed depreciated methods and fix test

File: tests/pipelines/pag/test_pag_sd_img2img.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+        expected_slice = np.array(
+            [0.5986328, 0.52441406, 0.3972168, 0.4741211, 0.34985352, 0.22705078, 0.4128418, 0.2866211, 0.31713867]
+        )
+        print(image_slice.flatten())
+        assert (
+            np.abs(image_slice.flatten() - expected_slice).max() < 1e-3
+        ), f"output is different from expected, {image_slice.flatten()}"
+
+@slow
+@require_torch_gpu
```
Comment (yiyixuxu) [inline tests/pipelines/pag/test_pag_sd_img2img.py:274]: this test is not needed? (same as the one above it?)

**Review by yiyixuxu — COMMENTED** [review]
thanks! I think we still have one test needs to be removed, and then can merge!

File: tests/pipelines/pag/test_pag_sd_img2img.py (hunk 2 above)
Comment (SahilCarterr) [reply tests/pipelines/pag/test_pag_sd_img2img.py]: should i remove `StableDiffusionPAGImg2ImgPipelineIntegrationTests(unittest.TestCase):`  also ?

File: src/diffusers/pipelines/pag/pipeline_pag_sd_img2img.py [hunk 3]
```diff
@@ -0,0 +1,1090 @@
...
+        """
+
+        if isinstance(callback_on_step_end, (PipelineCallback, MultiPipelineCallbacks)):
+            callback_on_step_end_tensor_inputs = callback_on_step_end.tensor_inputs
+
+        # 1. Check inputs. Raise error if not correct
+        self.check_inputs(
+            prompt,
+            strength,
+            callback_steps,
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/pipeline_pag_sd_img2img.py:898]:
```suggestion
```

**Comment by yiyixuxu** [comment]
@SahilCarterr can you look into failing tests?

**Comment by SahilCarterr** [comment]
Screenshot 2024-09-25 at 10 47 58 PM

3 errors remaining to be fix

**Comment by SahilCarterr** [comment]
Help Needed to fix the remaining errors

**Comment by SahilCarterr** [comment]
The errors are now fixed @yiyixuxu 
Screenshot 2024-10-06 at 1 25 56 PM

**Comment by yiyixuxu** [comment]
can you run `make style` and `make fix-copies`?

**Comment by SahilCarterr** [comment]
fixed make error

**Review by yiyixuxu — APPROVED** [review]
thank you!

**Pull request merged**